go/kbfs/libpages/stats_mysql.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2020 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libpages
+
+import (
+	"context"
+	"database/sql"
+	"sync"
+	"time"
+
+	"github.com/keybase/client/go/kbfs/data"
+	"github.com/keybase/client/go/kbfs/libkbfs"
+	"github.com/keybase/client/go/kbfs/tlf"
+	"go.uber.org/zap"
+)
+
+type mysqlActivityStatsStorer struct {
+	logger *zap.Logger
+	db     *sql.DB
+	clock  libkbfs.Clock
+
+	lock  sync.RWMutex
+	tlfs  map[tlf.ID]time.Time
+	hosts map[string]time.Time
+}
+
+func newMySQLActivityStatsStorerNoStart(clock libkbfs.Clock,
+	db *sql.DB, logger *zap.Logger) *mysqlActivityStatsStorer {
+	return &mysqlActivityStatsStorer{
+		logger: logger,
+		db:     db,
+		clock:  clock,
+		tlfs:   make(map[tlf.ID]time.Time),
+		hosts:  make(map[string]time.Time),
+	}
+}
+
+// NewMySQLActivityStatsStorer creates an ActivityStatsStorer that stores
+// activities on a MySQL database.
+func NewMySQLActivityStatsStorer(
+	db *sql.DB, logger *zap.Logger) ActivityStatsStorer {
+	s := newMySQLActivityStatsStorerNoStart(data.WallClock{}, db, logger)
+	// TODO shutdown()
+	go s.insertLoop(context.Background())
+	return s
+}
+
+func (s *mysqlActivityStatsStorer) createTablesIfNotExists(
+	ctx context.Context) (err error) {
+	if _, err = s.db.ExecContext(ctx, `
+        CREATE TABLE stats_tlf (
+          id          bigint unsigned NOT NULL AUTO_INCREMENT,
+          tlf_id      char(32)        NOT NULL,
+          active_time datetime(3)     NOT NULL,
+
+          PRIMARY KEY                 (id),
+          UNIQUE KEY  idx_tlf_id      (tlf_id),
+          KEY         idx_active_time (active_time)
+        ) ENGINE=InnoDB DEFAULT CHARSET=utf8
+    `); err != nil {
+		return err
+	}
+	if _, err = s.db.ExecContext(ctx, `
+        CREATE TABLE stats_host (
+          id          bigint unsigned NOT NULL AUTO_INCREMENT,
+          domain      varchar(256)    NOT NULL,
+          active_time datetime(3)     NOT NULL,
+
+          PRIMARY KEY                 (id),
+          UNIQUE KEY  idx_domain      (domain),
+          KEY         idx_active_time (active_time)
+        ) ENGINE=InnoDB DEFAULT CHARSET=utf8
+    `); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+const mysqlStatFlushTimeout = time.Minute / 2
+
+func (s *mysqlActivityStatsStorer) flushInserts() {
+	s.lock.RLock()
```

Comment:
Since this block updates `s.tlfs` and `s.hosts`, a write lock seems more appropriate.  Since only one of these should be happening at once anyway it shouldn't matter, but I think a full write lock would make that more clear.